### ISSUE TITLE
Syntax error: "(" unexpected (expecting "fi")

### ISSUE THREAD

**Issue by ulysses4ever** [mchav/with#15]
I downloaded the script, run it on my Ubuntu 16.04 laptop and get

```
$ with ls
/home/ulysses/bin/with: 12: /home/ulysses/bin/with: Syntax error: "(" unexpected (expecting "fi")
```

**Comment by eush77** [comment]
Same error for me under Dash (which is what /bin/sh is sym-linked to on my system). Runs ok under Bash though.

**Comment by ulysses4ever** [comment]
I tried it (and got above mentioned error) in `gnome-terminal`.

**Comment by ulysses4ever** [comment]
Oh, I see: I also have `/bin/sh` synlinked to `dash`. I just changed the script to call `bash`. Though it would be nice if authors accounted this somehow.

**Comment by eush77** [comment]
Yeah, the workaround is to change the interpreter to Bash. Although it was changed the other way in 61d059d so it is probably in the author's intentions to make it run under any POSIX shell.

Either way, I don't think this issue should be closed right now.

**Comment by mchav** [comment]
Yes @eush77 I'd like it to run under any Posix shell. We'll keep the issue open. If anyone has any insights on how to run it on dash please let me know.

**Comment by eush77** [comment]
I ran [shellcheck](http://www.shellcheck.net/) on this script with the first line changed to `#!/bin/sh` and it found several Bash-specific features used by the code:
- arrays;
- here-strings;
- `read -e` (readline mode);
- escaped string literals (`$'\n'`);
- `declare` and `type` builtins.

And it somehow missed this:
- substring expansion (`${parameter:offset:length}`)— [not specified](http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_02) by POSIX.

Overall, it seems highly non-trivial to make `with` run under any POSIX shell, which lack some crucial features that are otherwise supported by Bash (how can we work around missing `read -e`?).
